parser_library/src/context/ordinary_assembly/address.cpp

```diff
@@ -77,177 +82,282 @@ void space::resolve(space_ptr this_space, space_ptr value)
+    std::vector<address::base_entry> result;
 
-        res.push_back(*rest);
-        if (operation == op::SUB)
-            res.back().second *= -1;
+    result.reserve(l.size() + r.size());
+
+    result = r;
+    if constexpr (operation == merge_op::sub)
+    {
+        for (auto& [_, cnt] : result)
+            cnt *= -1;
     }
 
-    return res;
+    std::sort(result.begin(), result.end(), [](const auto& l, const auto& r) { return l.first < r.first; });
+    utils::merge_unsorted(
+        result,
+        l,
+        [](const auto& l, const auto& r) { return l.first <=> r.first; },
+        [](auto& r, const auto& e) { r.second += e.second; });
+
+    std::erase_if(result, [](const auto& e) { return e.second == 0; });
+
+    return result;
 }
 
 address address::operator+(const address& addr) const
 {
-    return address(merge_entries(bases_, addr.bases_, op::ADD),
-        offset() + addr.offset(),
-        merge_entries(normalized_spaces(), addr.normalized_spaces(), op::ADD));
+    if (!has_spaces() && !addr.has_spaces())
+        return address(merge_bases<merge_op::add>(bases_, addr.bases_), offset_ + addr.offset_, space_list());
+
+    auto res_spaces = std::make_shared<std::vector<space_entry>>();
+    normalization_helper helper;
+
+    int offset = 0;
+    offset += get_unresolved_spaces(spaces_.spaces, helper, *res_spaces, 1);
+    offset += get_unresolved_spaces(addr.spaces_.spaces, helper, *res_spaces, 1);
+
+    cleanup_spaces(*res_spaces);
+
+    return address(merge_bases<merge_op::add>(bases_, addr.bases_),
+        offset_ + addr.offset_ + offset,
+        space_list(std::move(res_spaces)));
 }
 
 address address::operator+(int offs) const { return address(bases_, offset_ + offs, spaces_); }
 
+std::pair<std::span<const address::space_entry>, std::span<const address::space_entry>> trim_common(
+    std::span<const address::space_entry> l, std::span<const address::space_entry> r)
+{
+    if (l.data() == r.data())
+    {
+        auto common = std::min(l.size(), r.size());
+        return { l.subspan(common), r.subspan(common) };
+    }
+    auto [le, re] = std::mismatch(l.begin(), l.end(), r.begin(), r.end());
+    return { { le, l.end() }, { re, r.end() } };
+}
+
 address address::operator-(const address& addr) const
 {
-    return address(merge_entries(bases_, addr.bases_, op::SUB),
-        offset() - addr.offset(),
-        merge_entries(normalized_spaces(), addr.normalized_spaces(), op::SUB));
+    auto [lspaces, rspaces] = trim_common(spaces_.spaces, addr.spaces_.spaces);
+    if (lspaces.empty() && rspaces.empty())
+        return address(merge_bases<merge_op::sub>(bases_, addr.bases_), offset_ - addr.offset_, {});
+
+    normalization_helper helper;
+
+    size_t l_processed = 0;
+
+    if (rspaces.empty())
+    {
+        for (const auto& [sp, cnt] : lspaces)
+        {
+            if (cnt == 0 || sp->resolved()
+                || !helper.map.try_emplace(sp.get(), l_processed).second) // requires normalization
```

Comment:
Just to be sure - is this a `todo` or a regular note?

Reply:
This is just a note why we break the loop.